rosidl_typesupport_opensplice_cpp/resource/msg__type_support.cpp.em

```diff
@@ -122,55 +142,57 @@ register_type__@(spec.base_type.type)(
   }
 }
 
-ROSIDL_TYPESUPPORT_OPENSPLICE_CPP_EXPORT_@(spec.base_type.pkg_name)
+ROSIDL_TYPESUPPORT_OPENSPLICE_CPP_EXPORT_@(package_name)
 void
 convert_ros_message_to_dds(const __ros_msg_type & ros_message, __dds_msg_type & dds_message)
 {
-@[if not spec.fields]@
+@[if not message.structure.members]@
   (void)ros_message;
   (void)dds_message;
 @[end if]@
-@[for field in spec.fields]@
-  // field.name @(field.name)
-@[  if field.type.is_array]@
+@[for member in message.structure.members]@
+  // member.name @(member.name)
+@[  if isinstance(member.type, NestedType)]@
   {
-@[    if field.type.array_size and not field.type.is_upper_bound]@
-    size_t size = @(field.type.array_size);
+@[    if isinstance(member.type, Array)]@
+    size_t size = @(member.type.size);
 @[    else]@
-    size_t size = ros_message.@(field.name).size();
+    size_t size = ros_message.@(member.name).size();
+@[    end if]@
     if (size > (std::numeric_limits<DDS::Long>::max)()) {
       throw std::runtime_error("array size exceeds maximum DDS sequence size");
     }
     DDS::Long length = static_cast<DDS::Long>(size);
-    dds_message.@(field.name)_.length(length);
-@[    end if]@
+    dds_message.@(member.name)_.length(length);
     for (DDS::ULong i = 0; i < size; i++) {
-@[    if field.type.type == 'string']@
-      dds_message.@(field.name)_[i] = ros_message.@(field.name)[i].c_str();
-@[    elif field.type.type == 'bool']@
-      dds_message.@(field.name)_[i] = 1 ? ros_message.@(field.name)[i] : 0;
-@[    elif field.type.is_primitive_type()]@
-      dds_message.@(field.name)_[i] = ros_message.@(field.name)[i];
+@[    if isinstance(member.type.basetype, BaseString)]@
+      dds_message.@(member.name)_[i] = ros_message.@(member.name)[i].c_str();
```

Comment:
You're probably right, unless the ROS message is using a std::wstring. How is a WString represented in the ROS message and what should we do for the conversion?

Reply:
We don't have any examples / tests for that yet. The representation of WString in each language is described here: https://github.com/ros2/design/blob/gh-pages/articles/142_idl.md#type-mapping

Reply:
Ok, since `WString` is described as having type `std::u16string`, I think the use of `c_str()` should work since it is a [std::basic_string](https://en.cppreference.com/w/cpp/string/basic_string).